mypy/checker.py

```diff
@@ -1676,6 +1677,57 @@ def visit_class_def(self, defn: ClassDef) -> None:
         if typ.is_protocol and typ.defn.type_vars:
             self.check_protocol_variance(defn)
 
+    def check_init_subclass(self, defn: ClassDef) -> None:
+        """
+        Check that the number of args to __init_subclass__, as well as typing are  both correct
+        In this example:
+
+            1   class Base:
+            2       def __init_subclass__(cls, thing: int):
+            3           pass
+            4   class Child(Base, thing=5):
+            5       def __init_subclass__(cls):
+            6           pass
+            7   Child()
+
+        Base.__init_subclass__(thing=5) is called at line 4. This is what we simulate here
+        Child.__init_subclass__ is never called
+        """
+        typ = defn.info
+        # at runtime, only Base.__init_subclass__ will be called
+        # we skip the current class itself
+        for base in typ.mro[1:]:
+            # 'object.__init_subclass__ is a dummy method with no arguments, always defined
+            # there is no use to call it
+            if base.name() != 'object' \
+                    and base.defn.info:  # there are "NOT_READY" instances
+                # during the tests, so I filter them out...
+                for method_name, method_symbol_node in base.defn.info.names.items():
+                    if method_name == '__init_subclass__':
+                        name_expr = NameExpr(defn.name)
+                        name_expr.node = base
+                        callee = MemberExpr(name_expr, '__init_subclass__')
+                        args = list(defn.keywords.values())
+                        arg_names = list(defn.keywords.keys())  # type: List[Optional[str]]
+                        # 'metaclass' keyword is consumed by the rest of the type machinery,
+                        # and is never passed to __init_subclass__ implementations
+                        if 'metaclass' in arg_names:
+                            idx = arg_names.index('metaclass')
+                            arg_names.pop(idx)
+                            args.pop(idx)
+                        arg_kinds = [ARG_NAMED] * len(args)
+                        call_expr = CallExpr(callee, args, arg_kinds, arg_names)
+                        call_expr.line = defn.line
+                        call_expr.column = defn.column
+                        call_expr.end_line = defn.end_line
+                        self.expr_checker.accept(call_expr,
```

Comment:
Using `accept()` may actually be dangerous. For example if the base class is not in scope (defined in another module and not imported in the current one), this may trigger `Name not defined` or maybe even a crash. There two possible solutions here:
* Point `name_expr.node` to the base class `TypeInfo` before calling `accept()`
* Use `check_call()` or similar methods

Please add a test for this.

Reply:
Hm sorry, I don't understand. I tried to produce a mypy runtime exception with 2 modules, one of  which defines a _Base_, and the other one **doesn't** import that _Base_, and it only gave me a nice error `Name 'Base' is not defined`. Which os OK I guess? this is what would happen at runtime anyway (`NameError: name 'Base' is not defined`)

Reply:
> Which os OK I guess?

It is not what I wanted. I wanted something like this:
```python
# file bases.py
class Base:
    def __init_subclass__(cls, **kwargs) -> None: ...
class MidBase(Base):
    ...
# file main.py
from bases import MidBase
class Main(MidBase, test=False):
    ...
```